synapse/lib/kv.py

```diff
@@ -0,0 +1,325 @@
+import os
+import lmdb
+import functools
+
+import synapse.common as s_common
+import synapse.eventbus as s_eventbus
+
+import synapse.lib.msgpack as s_msgpack
+import synapse.lib.threads as s_threads
+
+class KvDict:
+    '''
+    A KvDict uses the KvStor to implement a pythonic dict-like object.
+    '''
+    def __init__(self, stor, iden):
+        self.stor = stor
+        self.iden = iden
+        self.vals = {}
+
+        # load all the props
+        for lkey, lval in stor.iterKvProps(iden):
+            prop = lkey[16:].decode('utf8')
+            self.vals[prop] = s_msgpack.un(lval)
+
+    def set(self, prop, valu):
+        '''
+        Set a property in the KvDict.
+
+        Args:
+            prop (str): The property name.
+            valu (obj): A msgpack compatible value.
+        '''
+        self.vals[prop] = valu
+
+        lkey = self.iden + prop.encode('utf8')
+        self.stor.setKvProp(lkey, s_msgpack.en(valu))
+
+    def get(self, prop):
+        '''
+        Get a property from the KvDict.
+
+        Args:
+            prop (str): The property name.
+
+        Returns:
+            (obj): The return value, or None.
+        '''
+        return self.vals.get(prop)
+
+class KvLook:
+    '''
+    A KvLook uses the KvStor to implement key=valu lookup namespace.
+    '''
+    def __init__(self, stor, iden):
+        self.stor = stor
+        self.iden = iden
+
+    def set(self, prop, valu):
+        '''
+        Set a property in the KvLook.
+
+        Args:
+            prop (str): The property name.
+            valu (obj): A msgpack compatible value.
+        '''
+        lkey = self.iden + prop.encode('utf8')
+        self.stor.setKvProp(lkey, s_msgpack.en(valu))
+
+    def get(self, prop):
+        '''
+        Lookup a property from the KvLook.
+
+        Args:
+            prop (str): The property name.
+
+        Returns:
+            (obj): The return value, or None.
+        '''
+        lkey = self.iden + prop.encode('utf8')
+
+        lval = self.stor.getKvProp(lkey)
+        if lval is None:
+            return None
+
+        return s_msgpack.un(lval)
+
+class KvList:
+    '''
+    A KvList uses the KvStor to implement a pythonic list-like object.
+    '''
+    def __init__(self, stor, iden):
+        self.stor = stor
+        self.iden = iden
+
+        self.vals = [s_msgpack.un(b) for b in stor.iterKvDups(iden)]
+
+    def remove(self, valu):
+        '''
+        Remove a value from the KvList.
+
+        Args:
+            valu (obj): A msgpack value to remove.
+
+        Returns:
+            (bool): True if the value was found and removed.
+        '''
+        try:
+            self.vals.remove(valu)
+        except ValueError as e:
+            return False
+
+        lval = s_msgpack.en(valu)
+        self.stor.delKvDup(self.iden, lval)
+        return True
+
+    def __iter__(self):
+        return iter(self.vals)
+
+    def __len__(self):
+        return len(self.vals)
+
+    def append(self, valu):
```

Comment:
For the methods on this class, do we need to wrap the parts that modify `vals` and the parts that modify the `stor` in a lock?

Reply:
We certainly could.  The only race we'd really be fighting is an issue of ordering. Since we dont currently ( and have no intention ) to support order oriented operations like insert(), we could also just say that the list is inherently un-ordered ( from a persistence perspective ) and bask in the glory of lock-free operation.  ;)  Thoughts?

Reply:
@vertexmc @invisig0th if locking is still a question, please comment on #628